### PR TITLE
BREAKING CHANGE: Deno.Reader is async iterator, remove toAsyncIterator

### LINKED ISSUE #4831: Deno.Reader should be AsyncIterable
Ideally this would be done without having every Reader implementor also implementing `[Symbol.iterator]()`, only the `read()` method

Probably Deno.Reader needs to become an abstract base class instead of interface.

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes #4831

### CODE REVIEW

File: cli/js/io.ts [hunk 1]
```diff
@@ -14,10 +14,30 @@ export enum SeekMode {
...
+    const result = await this.read(this.#b);
+    if (result === EOF) {
+      return { value: new Uint8Array(), done: true };
+    }
+
+    return {
+      value: this.#b.subarray(0, result),
+      done: false,
+    };
+  }
```
Comment (nayeemrmn) [inline cli/js/io.ts:40]: Generator? You can avoid exposing a `next()` method.

Comment (bartlomieju) [reply cli/js/io.ts:40]: Good idea, thanks

**Comment by ry** [comment]
I asked for this change, but after discussing with @bartlomieju I'm not sure this is better.

**Comment by bartlomieju** [comment]
I'm not really in favor of this change; JS supports only single `extends` - sooner or later we'd need to support [`mixin` pattern for classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#Mix-ins) for downstream consumers. 

Not sure if that code in the runtime would be a net gain. 

This PR has single use: `for await (const chunk of <Deno.Reader instance>)`; makes it harder to mock out `Deno.Reader`; and in my personal opinion isn't more readable than `for await (const chunk of Deno.toAsyncIterator(<Deno.Reader instance>))`.

**Comment by bartlomieju** [comment]
Discussed offline; we're not gonna pursue this change.

**Comment by alisabzevari** [comment]
Wouldn't it make sense to still change `Reader` interface to `extend AsyncIterable<Uint8Array>`? All the implementors could call `Deno.iter` to implement it and it would be more convenient for `Reader` users.

**Comment by ry** [comment]
@alisabzevari It doesn't seem to correct to require implementors to all implement the same method.

**Pull request closed without merging**